daemon/p2p_announce.c

```diff
@@ -0,0 +1,322 @@
+#include "daemon/chaintopology.h"
+#include "daemon/log.h"
+#include "daemon/p2p_announce.h"
+#include "daemon/packets.h"
+#include "daemon/peer.h"
+#include "daemon/routing.h"
+#include "daemon/secrets.h"
+#include "daemon/timeout.h"
+
+#include <arpa/inet.h>
+#include <ccan/tal/str/str.h>
+#include <ccan/tal/tal.h>
+#include <secp256k1.h>
+
+
+u8 ipv4prefix[] = {
+	0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0xFF, 0xFF
+};
+
+/* Read an IP from `srcip` and convert it into the dotted
+ * notation. Handles both IPv4 and IPv6 addresses and converts
+ * accordingly. We differentiate the two by using the RFC 4291
+ * IPv4-mapped IPv6 format */
+static char* read_ip(const tal_t *ctx, const struct ipv6 *srcip)
+{
+	char tempaddr[INET6_ADDRSTRLEN];
+
+	if (memcmp(srcip, ipv4prefix, sizeof(ipv4prefix)) == 0) {
+		inet_ntop(AF_INET, srcip + 12, tempaddr, sizeof(tempaddr));
+	}else{
+		inet_ntop(AF_INET6, srcip, tempaddr, sizeof(tempaddr));
+	}
+	return tal_strdup(ctx, tempaddr);
+}
+
+/* Serialize the IP address in `srcip` into a 16 byte
+ * representation. It handles both IPv6 and IPv4 addresses, prefixing
+ * IPv4 addresses with the prefix described in RFC 4291. */
+static void write_ip(struct ipv6 *dstip, char *srcip)
+{
+	if (!strchr(srcip, ':')) {
+		memcpy(dstip, ipv4prefix, sizeof(ipv4prefix));
+		inet_pton(AF_INET, srcip, dstip);
+	} else {
+		inet_pton(AF_INET6, srcip, dstip);
+	}
+}
+
+static void broadcast(struct lightningd_state *dstate,
+		      int type, u8 *pkt,
+		      struct peer *origin)
+{
+	struct peer *p;
+	list_for_each(&dstate->peers, p, list) {
+		if (state_is_normal(p->state) && origin != p)
+			queue_pkt_nested(p, type, pkt);
+	}
+}
+
+static bool add_channel_direction(struct lightningd_state *dstate,
+				  const struct pubkey *from,
+				  const struct pubkey *to,
+				  const int direction,
+				  const struct channel_id *channel_id
+	)
+{
+	struct node_connection *c = get_connection(dstate, from, to);
+	if (c){
+		/* Do not clobber connections added otherwise */
+		memcpy(&c->channel_id, channel_id, sizeof(c->channel_id));
+		c->flags = direction;
+		printf("Found node_connection via get_connection");
+		return false;
+	}else if(get_connection_by_cid(dstate, channel_id, direction)) {
+		return false;
+	}
+	half_add_connection(dstate,
+			    from,
+			    to,
+			    channel_id, direction);
+	return true;
+}
+
+void handle_channel_announcement(
+	struct peer *peer,
+	const struct msg_channel_announcement *msg)
+{
+	u8 *serialized;
+	bool forward = false;
+	if (!msg)
```

Comment:
Yeah, I was hoping to push this into the boiler plate code further up the call-tree with the `struct wire_message` so that we could unify the handling for parse failures, instead of having a null-check in every switch case for specific types. This will get removed once I merge that in.

`msg` can be null if the parser encounters a malformed packet.